File: Leetcode Daily Challenge/025-July-2025/026-maximize-subarrays-after-removing-one-conflicting-pair.java
// https://leetcode.com/problems/maximize-subarrays-after-removing-one-conflicting-pair/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(N)
    public long maxSubarrays(int n, int[][] conflictingPairs) {
        List<List<Integer>> conflictPoints = new ArrayList<>();
        for (int i = 0; i <= n; i ++) {
            conflictPoints.add(new ArrayList<>());
        }

        for (int[] pair : conflictingPairs) {
            if (pair[0] > pair[1]) conflictPoints.get(pair[0]).add(pair[1]);
            else conflictPoints.get(pair[1]).add(pair[0]);
        }

        long[] extraPoints = new long[n + 1];
        // extra[i] = no. of extra subarray after removing conflictingPair i

        long totalValidSubarray = 0, firstMaxConflict = 0, secondMaxConflict = 0;

        for (int i = 1; i <= n; i ++) { // total valid subarrays at index i

            for (int val : conflictPoints.get(i)) {
                if (val >= firstMaxConflict) {
                    secondMaxConflict = firstMaxConflict;
                    firstMaxConflict  = val;
                } else if (val > secondMaxConflict) {
                    secondMaxConflict = val;
                }
            }

            totalValidSubarray += (i - firstMaxConflict);
            extraPoints[(int)firstMaxConflict] += (firstMaxConflict - secondMaxConflict); 
        }

        long extraSubarray = Arrays.stream(extraPoints).max().getAsLong();
        return (totalValidSubarray + extraSubarray);
    }
}

File: Leetcode Daily Challenge/025-July-2025/026-maximize-subarrays-after-removing-one-conflicting-pair.cpp
// https://leetcode.com/problems/maximize-subarrays-after-removing-one-conflicting-pair/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    using ll = long long;

    // TC : O(N)
    // SC : O(N)
    long long maxSubarrays(int n, vector<vector<int>>& conflictingPairs) {
        unordered_map<int, vector<int>> mpp;    // stores the conflictingPoints for each number (a, b) where a <= b
        
        for (vector<int> &cp : conflictingPairs) {
            if (cp[0] > cp[1]) swap(cp[0], cp[1]);
            mpp[cp[1]].push_back(cp[0]);
        }

        // extra[i] = number of extra subarray on removal of conflicting pair i
        vector<ll> extraPoints(n + 1, 0); 

        ll totalValidSubarrays = 0, firstMaxConfict = 0, secondMaxConflict = 0;

        for (int i = 1; i <= n; i ++) { // subarray ending at i
            
            for (int &val : mpp[i]) {
                if (val >= firstMaxConfict) {
                    secondMaxConflict = firstMaxConfict;
                    firstMaxConfict   = val;
                } else if (val > secondMaxConflict) {
                    secondMaxConflict = val;
                }
            }

            totalValidSubarrays += (i - firstMaxConfict);

            extraPoints[firstMaxConfict] += (firstMaxConfict - secondMaxConflict);
        }

        return (totalValidSubarrays + *max_element(begin(extraPoints), end(extraPoints)));
    }
};